File: android/app/src/main/kotlin/com/stockbook/app/feature/book/ExpenseSummarySheet.kt
package com.stockbook.app.feature.book

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.stockbook.app.design.FadedRule
import com.stockbook.app.design.Icon
import com.stockbook.app.design.IconButton
import com.stockbook.app.design.Metrics
import com.stockbook.app.design.Nocturne
import com.stockbook.app.design.NocturneType
import com.stockbook.app.design.SheetHeader
import com.stockbook.app.design.card
import com.stockbook.app.design.hairline
import com.stockbook.core.model.ShopState
import com.stockbook.core.model.StatementPeriod
import com.stockbook.core.store.StockbookStore
import com.stockbook.core.text.Strings
import com.stockbook.core.text.SummaryDocument
import java.time.Instant
import java.time.ZoneId

/**
 * Where a month's money went, folded by what it went on.
 *
 * **The other question about the same expenses.** The list behind this sheet
 * answers "what did I spend", one receipt a line, over whatever span the picker
 * says. This answers "where did the month go" — and the two are not the same
 * page in two styles: a register is checked against the receipts in a drawer
 * line by line, and a folded page cannot be checked against anything. Which is
 * why the wording keeps them apart, `Report` against `Summary`, everywhere they
 * are named.
 *
 * **A month at a time, and only a month.** The book's own picker offers a year
 * and a hand-picked stretch as well; neither belongs here. "Petrol, 84 times" is
 * a fact about a year that tells the owner nothing they can act on, and the
 * whole use of this page is comparing one month against the last — which is what
 * the two arrows are for.
 *
 * Everything drawn comes out of [SummaryDocument], so what is read on the screen
 * and what comes out of the printer cannot drift.
 */
@Composable
fun ExpenseSummarySheet(
    /** Any instant inside the month being folded. */
    month: Instant,
    state: ShopState,
    store: StockbookStore,
    strings: Strings,
    /** Steps to another month without closing the sheet. */
    onMonth: (Instant) -> Unit,
    /** Renders this month as a page and hands it to the chooser. */
    onSave: () -> Unit,
    onClose: () -> Unit
) {
    // Keyed on the whole state, not read off the store bare: every getter here is
    // a plain function over a StateFlow snapshot, so an expense written while
    // this is open would otherwise leave the page showing the figures it had when
    // it was first drawn.
    val page = remember(month, state) {
        SummaryDocument.forSpendingSummary(
            store.spendingIn(StatementPeriod.Month(month)), month, state.settings, strings
        )
    }

    // No stepping into next month. A month that has not started has nothing in
    // it, and an owner who lands there wonders whether the app has lost their
    // records rather than that they walked past the end of the year.
    val isThisMonth = remember(month) {
        // Compared as calendar months in the phone's own zone, which is the zone
        // `StatementPeriod.Month` resolves in. Comparing the instants instead
        // would compare them in UTC, and an owner in Riyadh looking at their
        // spending at half past two on the morning of the 1st would find next
        // month already reachable.
        val zone = ZoneId.systemDefault()
        val here = month.atZone(zone).toLocalDate().withDayOfMonth(1)
        !here.isBefore(Instant.now().atZone(zone).toLocalDate().withDayOfMonth(1))
    }

    Column(modifier = Modifier.fillMaxWidth()) {
        SheetHeader(
            title = page.title,
            onClose = onClose,
            // Only where there is a month to hand over. A page saying nothing
            // happened is a page nobody needs a copy of.
            onShare = if (page.isEmpty) null else onSave
        )

        // `‹ August 2026 ›` — the month reads as the thing being stepped through
        // rather than as a caption with two buttons parked beside it. The same
        // shape the day summary steps days with, because it is the same gesture.
        Row(
            verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier.fillMaxWidth().padding(bottom = 16.dp)
        ) {
            IconButton(
                Icon.stepBack,
                onClick = { onMonth(month.stepMonth(-1)) },
                contentDescription = strings.previousMonth,
                tint = Nocturne.accent
            )
            Text(
                page.asOf,
                style = NocturneType.rowPrimary,
                color = Nocturne.text,
                textAlign = TextAlign.Center,
                maxLines = 1,
                modifier = Modifier.weight(1f)
            )
            if (isThisMonth) {
                // Holds the month in the middle when there is no forward arrow to
                // draw. Without it this month's name sits off to the right and
                // every step back nudges it, which reads as a bug.
                Spacer(Modifier.size(Metrics.minimumTouchTarget))
            } else {
                IconButton(
                    Icon.stepForward,
                    onClick = { onMonth(month.stepMonth(1)) },
                    contentDescription = strings.nextMonth,
                    tint = Nocturne.accent
                )
            }
        }

        if (page.isEmpty) {
            Text(
                page.emptyLine,
                style = NocturneType.meta,
                color = Nocturne.neutral500,
                modifier = Modifier.padding(bottom = 8.dp)
            )
            return@Column
        }

        // A plain Column rather than a lazy list: the sheet already scrolls its
        // content, and a lazy list given an unbounded height measures to zero and
        // draws no rows at all. A shop with more things to spend on than fit here
        // has a bigger problem than the scroll.
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .card()
                .hairline(radius = Metrics.cardRadius)
                .padding(14.dp)
        ) {
            page.rows.forEachIndexed { index, row ->
                SpendRow(row)
                if (index < page.rows.lastIndex) Spacer(Modifier.height(Metrics.rowGap))
            }

            // Inside the card it totals, under a rule — and it is the same figure
            // the Expense card on the pane shows for the same month, which is
            // what `SummaryDocumentTests` pins.
            FadedRule(modifier = Modifier.padding(top = 10.dp, bottom = 8.dp), inset = 0.dp)
            Row(modifier = Modifier.fillMaxWidth()) {
                Text(
                    page.totalLabel,
                    style = NocturneType.meta,
                    color = Nocturne.neutral500,
                    modifier = Modifier.weight(1f)
                )
                Text(page.totalValue, style = NocturneType.rowPrimary, color = Nocturne.text)
            }
        }
        Spacer(Modifier.height(4.dp))
    }
}

/**
 * One thing the money went on, and how often.
 *
 * The count sits under the name rather than in a column of its own: a phone row
 * has room for a name and a figure, and squeezing a third column between them
 * takes the room from the one thing on the row that has to stay readable.
 */
@Composable
private fun SpendRow(row: SummaryDocument.Row) {
    Row(verticalAlignment = Alignment.Top, modifier = Modifier.fillMaxWidth()) {
        Column(modifier = Modifier.weight(1f)) {
            Text(
                row.name,
                style = NocturneType.rowPrimary,
                color = Nocturne.text,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis
            )
            row.count?.let {
                Text(it, style = NocturneType.meta, color = Nocturne.neutral500)
            }
        }
        Spacer(Modifier.size(10.dp))
        Text(row.amount, style = NocturneType.rowPrimary, color = Nocturne.text)
    }
}

/**
 * One month back or forward, as a calendar reads it.
 *
 * Through `LocalDate` and re-anchored at midday, for the two reasons the day
 * stepper does the same: `Instant` only supports exact durations and a month is
 * not one, and an exact shift across a clock change can land back in the month it
 * started from — an arrow that appears to do nothing.
 *
 * The first of the month rather than the same day of it, so stepping back from
 * the 31st does not skip February.
 */
private fun Instant.stepMonth(by: Long): Instant {
    val zone = ZoneId.systemDefault()
    return atZone(zone).toLocalDate()
        .withDayOfMonth(1)
        .plusMonths(by)
        .atTime(12, 0)
        .atZone(zone)
        .toInstant()
}

File: ios/Stockbook/Features/Book/ExpenseSummarySheet.swift
import SwiftUI

/// Where a month's money went, folded by what it went on.
///
/// **The other question about the same expenses.** The list behind this sheet
/// answers "what did I spend", one receipt a line, over whatever span the picker
/// says. This answers "where did the month go" — and the two are not the same
/// page in two styles: a register is checked against the receipts in a drawer
/// line by line, and a folded page cannot be checked against anything. Which is
/// why the wording keeps them apart, `Report` against `Summary`, everywhere they
/// are named.
///
/// **A month at a time, and only a month.** The book's own picker offers a year
/// and a hand-picked stretch as well; neither belongs here. "Petrol, 84 times"
/// is a fact about a year that tells the owner nothing they can act on, and the
/// whole use of this page is comparing one month against the last — which is
/// what the two arrows are for.
///
/// Everything drawn comes out of `SummaryDocument`, so what is read on the
/// screen and what comes out of the printer cannot drift.
struct ExpenseSummarySheet: View {
    @Environment(StockbookStore.self) private var store

    /// Any date inside the month being folded.
    let month: Date
    /// Steps to another month without closing the sheet.
    let onMonth: (Date) -> Void
    let onClose: () -> Void

    /// The rendered page, waiting for the share sheet. Rendered on the tap rather
    /// than when the view is built, because the month changes underneath it.
    @State private var file: StatementFile?

    /// Read straight off the store rather than snapshotted into `@State`: it is
    /// `@Observable`, so an expense written while this is open redraws the line
    /// it belongs to. Android has to key this on the shop state by hand.
    private var page: SummaryDocument {
        SummaryDocument.forSpendingSummary(
            lines: store.spendingIn(.month(month)),
            monthOf: month,
            settings: store.settings,
            strings: Loc
        )
    }

    /// No stepping into next month. A month that has not started has nothing in
    /// it, and an owner who lands there wonders whether the app has lost their
    /// records rather than that they walked past the end of the year.
    private var isThisMonth: Bool {
        Calendar.current.compare(month, to: .now, toGranularity: .month) != .orderedAscending
    }

    var body: some View {
        let document = page

        VStack(alignment: .leading, spacing: 0) {
            SheetHeader(
                title: document.title,
                // Only where there is a month to hand over. A page saying nothing
                // happened is a page nobody needs a copy of.
                onShare: document.isEmpty ? nil : save,
                onClose: onClose
            )

            // `‹ August 2026 ›` — the month reads as the thing being stepped
            // through rather than as a caption with two buttons parked beside it.
            // The same shape the day summary steps days with, because it is the
            // same gesture.
            HStack {
                Button { onMonth(step(-1)) } label: { Glyph(Icon.stepBack, size: 18) }
                    .buttonStyle(.iconOnly)
                    .foregroundStyle(Nocturne.accent)
                    .accessibilityLabel(Loc.previousMonth)

                Text(document.asOf)
                    .nocturneText(.rowPrimary)
                    .lineLimit(1)
                    .frame(maxWidth: .infinity)

                if isThisMonth {
                    // Holds the month in the middle when there is no forward arrow
                    // to draw. Without it this month's name sits off to the right
                    // and every step back nudges it, which reads as a bug.
                    Color.clear.frame(width: Metrics.minimumTouchTarget, height: Metrics.minimumTouchTarget)
                } else {
                    Button { onMonth(step(1)) } label: { Glyph(Icon.stepForward, size: 18) }
                        .buttonStyle(.iconOnly)
                        .foregroundStyle(Nocturne.accent)
                        .accessibilityLabel(Loc.nextMonth)
                }
            }
            .padding(.bottom, 16)

            if document.isEmpty {
                Text(document.emptyLine)
                    .nocturneText(.meta)
                    .padding(.bottom, 8)
            } else {
                // A plain stack rather than a lazy one: the sheet already scrolls
                // its content, and a second scroll nested in the first is the trap
                // that has emptied a list on this codebase before. A shop with
                // more things to spend on than fit here has a bigger problem than
                // the scroll.
                VStack(alignment: .leading, spacing: 0) {
                    VStack(spacing: Metrics.rowGap) {
                        ForEach(Array(document.rows.enumerated()), id: \.offset) { _, row in
                            SpendRow(row: row)
                        }
                    }

                    // Inside the card it totals, under a rule — and it is the same
                    // figure the Expense card on the pane shows for the same
                    // month, which is what `SummaryDocumentTests` pins.
                    FadedRule(inset: 0)
                        .padding(.top, 10)
                        .padding(.bottom, 8)

                    HStack {
                        Text(document.totalLabel).nocturneText(.meta)
                        Spacer(minLength: 8)
                        Text(document.totalValue).nocturneText(.rowPrimary)
                    }
                }
                .frame(maxWidth: .infinity, alignment: .leading)
                // The card treatment written out rather than through a
                // `cardBox()` helper: the two that exist are `private extension
                // View` in the sheets that use them, so they are file-scoped and
                // a third copy here would earn nothing over the two lines.
                .padding(14)
                .nocturneCard()
                .padding(.bottom, 4)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .sheet(item: $file) { ShareSheet(url: $0.url) }
    }

    /// One month back or forward, as a calendar reads it.
    ///
    /// Through `Calendar` rather than by adding thirty days, for the reason the
    /// day stepper goes through it too: months are not a fixed length, and a step
    /// that landed back in the month it started from would look like an arrow
    /// that does nothing.
    private func step(_ by: Int) -> Date {
        Calendar.current.date(byAdding: .month, value: by, to: month) ?? month
    }

    /// A failure leaves `file` nil and nothing opens, which is the honest outcome
    /// the other printed pages already settled on.
    private func save() {
        guard let url = try? SummaryPDF.write(
            page,
            // Named for the month it folds, not for today: two prints of August
            // are the same page, and a folder of these is read by their file
            // names.
            fileName: Loc.expenseSummaryFileName(month: Copy.fileMonth(month))
        ) else { return }
        file = StatementFile(url: url)
    }
}

/// One thing the money went on, and how often.
///
/// The count sits under the name rather than in a column of its own: a phone row
/// has room for a name and a figure, and squeezing a third column between them
/// takes the room from the one thing on the row that has to stay readable.
private struct SpendRow: View {
    let row: SummaryDocument.Row

    var body: some View {
        HStack(alignment: .top, spacing: 10) {
            VStack(alignment: .leading, spacing: 0) {
                Text(row.name)
                    .nocturneText(.rowPrimary)
                    .lineLimit(1)
                if let count = row.count {
                    Text(count).nocturneText(.meta)
                }
            }
            Spacer(minLength: 8)
            Text(row.amount).nocturneText(.rowPrimary)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }
}
